Software Developer Job Market Prediction With APIs

Edgar Acosta

Purpose: To provide a detailed report of the current job market(regarding AI and Data Analytics) and to explore why I cannot obtain a job straight out of university.

First, start with the API request

In [1]:
import requests
import pandas as pd

APP_ID = "af21817b"
APP_KEY = "688d8e59766961f63d5e0ef54a3558cd	"

url = "https://api.adzuna.com/v1/api/jobs/ca/search/1"
params = {
    "app_id": APP_ID,
    "app_key": APP_KEY,
    "what": "Software Developer",       # change to "Data Analyst" for those roles
    "results_per_page": 500,
    "content-type": "application/json"
}

resp = requests.get(url, params=params)
if resp.status_code == 200:
    data = resp.json()
    jobs = data.get("results", [])
    employment_software = pd.DataFrame(jobs)
    print(employment_software[["title", "company", "location", "created"]].head())
else:
    print("Error fetching data:", resp.status_code, resp.text)

                         title  \
0           Software Developer   
1  Embedded Software Developer   
2           Software Developer   
3           Software Developer   
4           Software Developer   

                                             company  \
0  {'__CLASS__': 'Adzuna::API::Response::Company'...   
1  {'__CLASS__': 'Adzuna::API::Response::Company'...   
2  {'__CLASS__': 'Adzuna::API::Response::Company'...   
3  {'display_name': 'VLink Inc', '__CLASS__': 'Ad...   
4  {'display_name': 'CLPS Global', '__CLASS__': '...   

                                            location               created  
0  {'__CLASS__': 'Adzuna::API::Response::Location...  2025-05-22T05:22:50Z  
1  {'__CLASS__': 'Adzuna::API::Response::Location...  2025-06-11T23:09:14Z  
2  {'__CLASS__': 'Adzuna::API::Response::Location...  2025-05-31T22:31:15Z  
3  {'__CLASS__': 'Adzuna::API::Response::Location...  2025-05-29T23:14:48Z  
4  {'area': ['Canada', 'Ontario', 'Toronto'], 'di...  2025-05-29T23:13:12

In [2]:
import psutil
print(psutil.__version__)

7.0.0


In [3]:
url = "https://api.adzuna.com/v1/api/jobs/ca/search/1"
params = {
    "app_id": APP_ID,
    "app_key": APP_KEY,
    "what": "Data Analyst",       # change to "Data Analyst" for those roles
    "results_per_page": 500,
    "content-type": "application/json"
}

resp = requests.get(url, params=params)
if resp.status_code == 200:
    data = resp.json()
    jobs = data.get("results", [])
    employment_data = pd.DataFrame(jobs)
    print(employment_data[["title", "company", "location", "created"]].head())
else:
    print("Error fetching data:", resp.status_code, resp.text)

                    title                                            company  \
0            Data Analyst  {'__CLASS__': 'Adzuna::API::Response::Company'...   
1            Data Analyst  {'__CLASS__': 'Adzuna::API::Response::Company'...   
2            Data Analyst  {'display_name': 'Tata Consultancy Services', ...   
3  Technical Data Analyst  {'__CLASS__': 'Adzuna::API::Response::Company'...   
4            Data Analyst  {'__CLASS__': 'Adzuna::API::Response::Company'...   

                                            location               created  
0  {'area': ['Canada', 'Ontario', 'Toronto'], 'di...  2025-06-15T03:29:02Z  
1  {'area': ['Canada', 'Ontario', 'Toronto'], 'di...  2025-05-20T17:15:34Z  
2  {'__CLASS__': 'Adzuna::API::Response::Location...  2025-04-20T22:56:29Z  
3  {'__CLASS__': 'Adzuna::API::Response::Location...  2025-05-31T22:30:19Z  
4  {'__CLASS__': 'Adzuna::API::Response::Location...  2025-05-30T23:28:35Z  


In [4]:
employment_software

salary_is_predicted                                              title  \
0                    0                                 Software Developer   
1                    0                        Embedded Software Developer   
2                    0                                 Software Developer   
3                    0                                 Software Developer   
4                    0                                 Software Developer   
5                    0                           Software developer (C++)   
6                    0                          Senior Software Developer   
7                    0                          Senior Software Developer   
8                    0                             C++ Software Developer   
9                    0                    Senior Software Developer – API   
10                   0                          Senior Software Developer   
11                   0                  Software Developer and IT Support   
12                   0                    Software Developer II, Frontend   
13                   0                  Software Developer - API – Senior   
14                   0  Software Developer NodeJS and TypeScript -Remo...   
15                   0        PHP Software Developer & Product Support L3   
16                   0  Senior Software Development Engineer in Test (...   
17                   0  Software Developer (Python) - Up to $220k CAD ...   
18                   0  Senior Software developer - ETL development us...   
19                   0                                 Software Developer   
20                   0                                 Software Developer   
21                   0                                 Software Developer   
22                   0                                 Software Developer   
23                   0                                 Software Developer   
24                   0                                 Software Developer   
25                   0                                 Software Developer   
26                   0                                 Software Developer   
27                   0                                 Software Developer   
28                   0                             Software Developer Job   
29                   0                                 Software Developer   
30                   0                                 Software Developer   
31                   0                                 Software Developer   
32                   0                                 Software Developer   
33                   0                                 Software Developer   
34                   0                                 Software Developer   
35                   0                                 Software Developer   
36                   0                                 Software Developer   
37                   0                                 Software Developer   
38                   0                             Software developer Job   
39                   0                                 Software Developer   
40                   0                                 Software Developer   
41                   0                                 Software Developer   
42                   0                                 Software Developer   
43                   0                                 Software Developer   
44                   0                                 Software Developer   
45                   0                                 Software Developer   
46                   0                                 Software Developer   
47                   0                                 Software Developer   
48                   0                                 Software Developer   
49                   0                                 Software Developer   

                                             location  \
0   {'__CLASS__': 

In [5]:
employment_data

salary_is_predicted                                           location  \
0                    0  {'area': ['Canada', 'Ontario', 'Toronto'], 'di...   
1                    0  {'area': ['Canada', 'Ontario', 'Toronto'], 'di...   
2                    0  {'__CLASS__': 'Adzuna::API::Response::Location...   
3                    0  {'__CLASS__': 'Adzuna::API::Response::Location...   
4                    0  {'__CLASS__': 'Adzuna::API::Response::Location...   
5                    0  {'__CLASS__': 'Adzuna::API::Response::Location...   
6                    0  {'__CLASS__': 'Adzuna::API::Response::Location...   
7                    0  {'area': ['Canada'], 'display_name': 'Canada',...   
8                    0  {'__CLASS__': 'Adzuna::API::Response::Location...   
9                    0  {'area': ['Canada', 'Ontario', 'Toronto'], 'di...   
10                   0  {'__CLASS__': 'Adzuna::API::Response::Location...   
11                   0  {'display_name': 'Greater Vancouver, British C...   
12                   0  {'__CLASS__': 'Adzuna::API::Response::Location...   
13                   0  {'__CLASS__': 'Adzuna::API::Response::Location...   
14                   0  {'__CLASS__': 'Adzuna::API::Response::Location...   
15                   0  {'area': ['Canada', 'Ontario', 'Ottawa region'...   
16                   0  {'display_name': 'Greater Vancouver, British C...   
17                   0  {'__CLASS__': 'Adzuna::API::Response::Location...   
18                   0  {'area': ['Canada', 'British Columbia', 'Great...   
19                   0  {'area': ['Canada'], 'display_name': 'Canada',...   
20                   0  {'display_name': 'Greater Vancouver, British C...   
21                   0  {'__CLASS__': 'Adzuna::API::Response::Location...   
22                   0  {'__CLASS__': 'Adzuna::API::Response::Location...   
23                   0  {'area': ['Canada', 'Nova Scotia', 'Halifax re...   
24                   0  {'area': ['Canada', 'Québec', 'Montréal', 'Ver...   
25                   0  {'area': ['Canada', 'Québec', 'Laval region', ...   
26                   0  {'area': ['Canada', 'Ontario', 'Toronto'], 'di...   
27                   0  {'area': ['Canada', 'Ontario', 'Toronto', 'Cit...   
28                   0  {'display_name': 'Windsor, Essex region', 'are...   
29                   0  {'display_name': 'Toronto, Ontario', 'area': [...   
30                   0  {'__CLASS__': 'Adzuna::API::Response::Location...   
31                   0  {'__CLASS__': 'Adzuna::API::Response::Location...   
32                   0  {'__CLASS__': 'Adzuna::API::Response::Location...   
33                   0  {'__CLASS__': 'Adzuna::API::Response::Location...   
34                   0  {'display_name': 'Vancouver, Greater Vancouver...   
35                   0  {'area': ['Canada', 'British Columbia', 'Great...   
36                   0  {'__CLASS__': 'Adzuna::API::Response::Location...   
37                   0  {'display_name': 'Mississauga, Peel region', '...   
38                   0  {'__CLASS__': 'Adzuna::API::Response::Location...   
39                   0  {'area': ['Canada', 'Ontario', 'Peel region', ...   
40                   0  {'area': ['Canada', 'Ontario', 'Ottawa region'...   
41                   0  {'__CLASS__': 'Adzuna::API::Response::Location...   
42                   0  {'area': ['Canada', 'Alberta', 'Calgary region...   
43                   0  {'__CLASS__': 'Adzuna::API::Response::Location...   
44                   0  {'area': ['Canada'], 'display_name': 'Canada',...   
45                   0  {'__CLASS__': 'Adzuna::API::Response::Location...   
46                   0  {'area': ['Canada'], 'display_name': 'Canada',...   
47                   0  {'area': ['Canada'], 'display_name': 'Canada',...   
48                   0  {'__CLASS__': 'Adzuna::API::Response::Location...   
49                   0  {'display_name': 'Canada', 'area': ['Canada'],...   

                                          description  \
0   Akkodis is cur

Let us first merge(concatenate) these two datasets into one as both look into employment in areas in which I am interested in. Any Data cleaning and querying and analysis can be done in a single table since they both share the same attributes.

In [6]:
to_concat = [employment_data, employment_software]
employment_all = pd.concat(to_concat)
employment_all

salary_is_predicted                                           location  \
0                    0  {'area': ['Canada', 'Ontario', 'Toronto'], 'di...   
1                    0  {'area': ['Canada', 'Ontario', 'Toronto'], 'di...   
2                    0  {'__CLASS__': 'Adzuna::API::Response::Location...   
3                    0  {'__CLASS__': 'Adzuna::API::Response::Location...   
4                    0  {'__CLASS__': 'Adzuna::API::Response::Location...   
..                 ...                                                ...   
45                   0  {'area': ['Canada', 'Nova Scotia', 'Halifax re...   
46                   0  {'__CLASS__': 'Adzuna::API::Response::Location...   
47                   0  {'area': ['Canada', 'Ontario', 'Middlesex', 'L...   
48                   0  {'area': ['Canada'], 'display_name': 'Canada',...   
49                   0  {'display_name': 'London, Middlesex', 'area': ...   

                                          description                   title  \
0   Akkodis is currently seeking a skilled Power B...            Data Analyst   
1   About us Aquanow is a trading and technology c...            Data Analyst   
2   Inclusion without Exception Tata Consultancy S...            Data Analyst   
3   Position: Technical Data Analyst Location: Mon...  Technical Data Analyst   
4   We are seeking an analytical and detail-orient...            Data Analyst   
..                                                ...                     ...   
45  Overview Fundmetric is looking for a Software ...      Software Developer   
46  Job Introduction Frequentis Group with headqua...      Software Developer   
47  Who are we? Proax Technologies is a leading Te...      Software Developer   
48  Description We are seeking a Software Develope...      Software Developer   
49  Job Description Job Description Salary: Who ar...      Software Developer   

                                                adref  \
0   eyJhbGciOiJIUzI1NiJ9.eyJzIjoiYUFXVmhFTk44Qkd3a...   
1   eyJhbGciOiJIUzI1NiJ9.eyJpIjoiNTIwODE3MDY5MCIsI...   
2   eyJhbGciOiJIUzI1NiJ9.eyJpIjoiNTE1NDkyNzQzNyIsI...   
3   eyJhbGciOiJIUzI1NiJ9.eyJpIjoiNTIyNTY5ODg3MCIsI...   
4   eyJhbGciOiJIUzI1NiJ9.eyJzIjoiYUFXVmhFTk44Qkd3a...   
..                                                ...   
45  eyJhbGciOiJIUzI1NiJ9.eyJpIjoiNTIwNTAyMDI2MCIsI...   
46  eyJhbGciOiJIUzI1NiJ9.eyJpIjoiNTIzODczNDc0NiIsI...   
47  eyJhbGciOiJIUzI1NiJ9.eyJpIjoiNDg3MDQ5NDYxOSIsI...   
48  eyJhbGciOiJIUzI1NiJ9.eyJpIjoiNTIxNDQ4Njc5MiIsI...   
49  eyJhbGciOiJIUzI1NiJ9.eyJzIjoiQmxEQmcwTk44Qkd3a...   

                                             category               created  \
0   {'tag': 'it-jobs', 'label': 'IT Jobs', '__CLAS...  2025-06-15T03:29:02Z   
1   {'label': 'IT Jobs', 'tag': 'it-jobs', '__CLAS...  2025-05-20T17:15:34Z   
2   {'__CLASS__': 'Adzuna::API::Response::Category...  2025-04-20T22:56:29Z   
3   {'__CLASS__': 'Adzuna::API::Response::Category...  2025-05-31T22:30:19Z   
4   {'tag': 'accounting-finance-jobs', 'label': 'A...  2025-05-30T23:28:35Z   
..                                                ...                   ...   
45  {'__CLASS__': 'Adzuna::API::Response::Category...  2025-05-18T10:13:46Z   
46  {'label': 'IT Jobs', 'tag': 'it-jobs', '__CLAS...  2025-06-07T08:50:55Z   
47  {'__CLASS__': 'Adzuna::API::Response::Category...  2024-09-20T07:53:38Z   
48  {'tag': 'it-jobs', 'label': 'IT Jobs', '__CLAS...  2025-05-24T10:22:58Z   
49  {'__CLASS__': 'Adzuna::API::Response::Category...  2025-04-24T15:20:44Z   

   contract_type                                            company  \
0       contract  {'__CLASS__': 'Adzuna::API::Response::Company'...   
1      permanent  {'__CLASS__': 'Adzuna::API::Response::Company'...   
2            NaN  {'display_name': 'Tata Consultancy Services', ...   
3       contract  {'__CLASS__': 'Adzuna::API::Response::Company'...   
4            NaN  {'__CLASS__': 'Adzuna::API::Response::Company'...   
..           ...                             

Given the dataset above, there is so much to clean up. Many columns such as Class, adref or URL will not be used as they do not give the information regarding employment that is needed. Many columns have 0s or NAs(Unless there were only a few NAs per column, then Machine learning would be used to replace NA values). However, longitude and latitude will be used as they will be displayed on ArcGis

In [7]:
employment_all = employment_all.drop(['category','adref','__CLASS__','redirect_url',  'id','salary_is_predicted','contract_type','salary_max','salary_min','contract_time'], axis=1)

In [8]:
employment_all

location  \
0   {'area': ['Canada', 'Ontario', 'Toronto'], 'di...   
1   {'area': ['Canada', 'Ontario', 'Toronto'], 'di...   
2   {'__CLASS__': 'Adzuna::API::Response::Location...   
3   {'__CLASS__': 'Adzuna::API::Response::Location...   
4   {'__CLASS__': 'Adzuna::API::Response::Location...   
..                                                ...   
45  {'area': ['Canada', 'Nova Scotia', 'Halifax re...   
46  {'__CLASS__': 'Adzuna::API::Response::Location...   
47  {'area': ['Canada', 'Ontario', 'Middlesex', 'L...   
48  {'area': ['Canada'], 'display_name': 'Canada',...   
49  {'display_name': 'London, Middlesex', 'area': ...   

                                          description                   title  \
0   Akkodis is currently seeking a skilled Power B...            Data Analyst   
1   About us Aquanow is a trading and technology c...            Data Analyst   
2   Inclusion without Exception Tata Consultancy S...            Data Analyst   
3   Position: Technical Data Analyst Location: Mon...  Technical Data Analyst   
4   We are seeking an analytical and detail-orient...            Data Analyst   
..                                                ...                     ...   
45  Overview Fundmetric is looking for a Software ...      Software Developer   
46  Job Introduction Frequentis Group with headqua...      Software Developer   
47  Who are we? Proax Technologies is a leading Te...      Software Developer   
48  Description We are seeking a Software Develope...      Software Developer   
49  Job Description Job Description Salary: Who ar...      Software Developer   

                 created                                            company  \
0   2025-06-15T03:29:02Z  {'__CLASS__': 'Adzuna::API::Response::Company'...   
1   2025-05-20T17:15:34Z  {'__CLASS__': 'Adzuna::API::Response::Company'...   
2   2025-04-20T22:56:29Z  {'display_name': 'Tata Consultancy Services', ...   
3   2025-05-31T22:30:19Z  {'__CLASS__': 'Adzuna::API::Response::Company'...   
4   2025-05-30T23:28:35Z  {'__CLASS__': 'Adzuna::API::Response::Company'...   
..                   ...                                                ...   
45  2025-05-18T10:13:46Z  {'display_name': 'Fundmetric', '__CLASS__': 'A...   
46  2025-06-07T08:50:55Z  {'__CLASS__': 'Adzuna::API::Response::Company'...   
47  2024-09-20T07:53:38Z  {'display_name': 'Proax Technologies', '__CLAS...   
48  2025-05-24T10:22:58Z  {'display_name': 'Nokia Global', '__CLASS__': ...   
49  2025-04-24T15:20:44Z  {'display_name': 'Proax Technologies', '__CLAS...   

     latitude  longitude  
0         NaN        NaN  
1   43.706102 -79.402726  
2         NaN        NaN  
3         NaN        NaN  
4         NaN        NaN  
..        ...        ...  
45  44.681555 -63.556363  
46  45.466091 -75.704464  
47  42.933380 -81.205550  
48        NaN        NaN  
49  42.981065 -81.248057  

[100 rows x 7 columns]

The issue with this dataset is that there are no columns for employment levels, years of experience, and education levels. However, location and date created will be analyzed, as well as job titles as these could provide some insights on employment, what is available, where it is available, and which company is providing the vacancy.

First, it is important to clean up the data, and make it readable, especially the location, and company. created will be reformatted to provide year, month, day only.

Another issue is with location where some locations are displayed as Country, Province, and City, while others are displayed as city, region, and then the area

In [9]:
print(employment_all['location'])
print(employment_all['company'].iloc[0])

0     {'area': ['Canada', 'Ontario', 'Toronto'], 'di...
1     {'area': ['Canada', 'Ontario', 'Toronto'], 'di...
2     {'__CLASS__': 'Adzuna::API::Response::Location...
3     {'__CLASS__': 'Adzuna::API::Response::Location...
4     {'__CLASS__': 'Adzuna::API::Response::Location...
                            ...                        
45    {'area': ['Canada', 'Nova Scotia', 'Halifax re...
46    {'__CLASS__': 'Adzuna::API::Response::Location...
47    {'area': ['Canada', 'Ontario', 'Middlesex', 'L...
48    {'area': ['Canada'], 'display_name': 'Canada',...
49    {'display_name': 'London, Middlesex', 'area': ...
Name: location, Length: 100, dtype: object
{'__CLASS__': 'Adzuna::API::Response::Company', 'display_name': 'Akkodis'}


In [10]:
import ast
employment_all['location'].head(10).apply(type)
def normalize_location(loc):
    if isinstance(loc, str):
        try:
            loc = ast.literal_eval(loc)
        except Exception:
            return {}

    if isinstance(loc, dict):
        if 'area' in loc:
            return {
                'country': loc['area'][0] if len(loc['area']) > 0 else None,
                'region': loc['area'][1] if len(loc['area']) > 1 else None,
                'city': loc['area'][2] if len(loc['area']) > 2 else None
            }
        elif 'display_name' in loc:
            return {'display_name': loc['display_name']}
    return {}

# Normalize and convert to DataFrame
location_cleaned = employment_all['location'].apply(normalize_location)
location_df = pd.DataFrame(location_cleaned.tolist())

# Reset indices
employment_all = employment_all.reset_index(drop=True)
location_df = location_df.reset_index(drop=True)

# Join cleaned location info
employment_all = pd.concat([employment_all.drop(columns=['location']), location_df], axis=1)

In [11]:
employment_all

description                   title  \
0   Akkodis is currently seeking a skilled Power B...            Data Analyst   
1   About us Aquanow is a trading and technology c...            Data Analyst   
2   Inclusion without Exception Tata Consultancy S...            Data Analyst   
3   Position: Technical Data Analyst Location: Mon...  Technical Data Analyst   
4   We are seeking an analytical and detail-orient...            Data Analyst   
..                                                ...                     ...   
95  Overview Fundmetric is looking for a Software ...      Software Developer   
96  Job Introduction Frequentis Group with headqua...      Software Developer   
97  Who are we? Proax Technologies is a leading Te...      Software Developer   
98  Description We are seeking a Software Develope...      Software Developer   
99  Job Description Job Description Salary: Who ar...      Software Developer   

                 created                                            company  \
0   2025-06-15T03:29:02Z  {'__CLASS__': 'Adzuna::API::Response::Company'...   
1   2025-05-20T17:15:34Z  {'__CLASS__': 'Adzuna::API::Response::Company'...   
2   2025-04-20T22:56:29Z  {'display_name': 'Tata Consultancy Services', ...   
3   2025-05-31T22:30:19Z  {'__CLASS__': 'Adzuna::API::Response::Company'...   
4   2025-05-30T23:28:35Z  {'__CLASS__': 'Adzuna::API::Response::Company'...   
..                   ...                                                ...   
95  2025-05-18T10:13:46Z  {'display_name': 'Fundmetric', '__CLASS__': 'A...   
96  2025-06-07T08:50:55Z  {'__CLASS__': 'Adzuna::API::Response::Company'...   
97  2024-09-20T07:53:38Z  {'display_name': 'Proax Technologies', '__CLAS...   
98  2025-05-24T10:22:58Z  {'display_name': 'Nokia Global', '__CLASS__': ...   
99  2025-04-24T15:20:44Z  {'display_name': 'Proax Technologies', '__CLAS...   

     latitude  longitude country       region            city  
0         NaN        NaN  Canada      Ontario         Toronto  
1   43.706102 -79.402726  Canada      Ontario         Toronto  
2         NaN        NaN  Canada      Ontario     Peel region  
3         NaN        NaN  Canada         None            None  
4         NaN        NaN  Canada         None            None  
..        ...        ...     ...          ...             ...  
95  44.681555 -63.556363  Canada  Nova Scotia  Halifax region  
96  45.466091 -75.704464  Canada       Québec       Outaouais  
97  42.933380 -81.205550  Canada      Ontario       Middlesex  
98        NaN        NaN  Canada         None            None  
99  42.981065 -81.248057  Canada      Ontario       Middlesex  

[100 rows x 9 columns]

In [12]:
def extract_company_name(company):
    if isinstance(company, str):
        try:
            company = ast.literal_eval(company)
        except Exception:
            return None
    if isinstance(company, dict):
        return company['display_name']
    return None

employment_all['company_name'] = employment_all['company'].apply(extract_company_name)
employment_all = employment_all.drop(columns=['company'])

In [13]:
employment_all

description                   title  \
0   Akkodis is currently seeking a skilled Power B...            Data Analyst   
1   About us Aquanow is a trading and technology c...            Data Analyst   
2   Inclusion without Exception Tata Consultancy S...            Data Analyst   
3   Position: Technical Data Analyst Location: Mon...  Technical Data Analyst   
4   We are seeking an analytical and detail-orient...            Data Analyst   
..                                                ...                     ...   
95  Overview Fundmetric is looking for a Software ...      Software Developer   
96  Job Introduction Frequentis Group with headqua...      Software Developer   
97  Who are we? Proax Technologies is a leading Te...      Software Developer   
98  Description We are seeking a Software Develope...      Software Developer   
99  Job Description Job Description Salary: Who ar...      Software Developer   

                 created   latitude  longitude country       region  \
0   2025-06-15T03:29:02Z        NaN        NaN  Canada      Ontario   
1   2025-05-20T17:15:34Z  43.706102 -79.402726  Canada      Ontario   
2   2025-04-20T22:56:29Z        NaN        NaN  Canada      Ontario   
3   2025-05-31T22:30:19Z        NaN        NaN  Canada         None   
4   2025-05-30T23:28:35Z        NaN        NaN  Canada         None   
..                   ...        ...        ...     ...          ...   
95  2025-05-18T10:13:46Z  44.681555 -63.556363  Canada  Nova Scotia   
96  2025-06-07T08:50:55Z  45.466091 -75.704464  Canada       Québec   
97  2024-09-20T07:53:38Z  42.933380 -81.205550  Canada      Ontario   
98  2025-05-24T10:22:58Z        NaN        NaN  Canada         None   
99  2025-04-24T15:20:44Z  42.981065 -81.248057  Canada      Ontario   

              city                   company_name  
0          Toronto                        Akkodis  
1          Toronto                        Aquanow  
2      Peel region      Tata Consultancy Services  
3             None                   iTech US Inc  
4             None  Critical Control Technologies  
..             ...                            ...  
95  Halifax region                     Fundmetric  
96       Outaouais                   Pivot + Edge  
97       Middlesex             Proax Technologies  
98            None                   Nokia Global  
99       Middlesex             Proax Technologies  

[100 rows x 9 columns]

Just for the purposes of data cleaning, even if it is not accurate, latitude and longitude NaN values will be replaced via Machine Learning(KNN)

In [14]:
from sklearn.impute import KNNImputer
from sklearn.preprocessing import LabelEncoder

employment_all['city_id'] = LabelEncoder().fit_transform(employment_all['city'].astype(str)) #City is a string, create a new column that would enumerate city

# Use these for imputation
impute_cols = ['latitude', 'longitude', 'city_id']
imputer = KNNImputer(n_neighbors=3)

# Impute
employment_all[impute_cols] = imputer.fit_transform(employment_all[impute_cols])

In [15]:
employment_all

description                   title  \
0   Akkodis is currently seeking a skilled Power B...            Data Analyst   
1   About us Aquanow is a trading and technology c...            Data Analyst   
2   Inclusion without Exception Tata Consultancy S...            Data Analyst   
3   Position: Technical Data Analyst Location: Mon...  Technical Data Analyst   
4   We are seeking an analytical and detail-orient...            Data Analyst   
..                                                ...                     ...   
95  Overview Fundmetric is looking for a Software ...      Software Developer   
96  Job Introduction Frequentis Group with headqua...      Software Developer   
97  Who are we? Proax Technologies is a leading Te...      Software Developer   
98  Description We are seeking a Software Develope...      Software Developer   
99  Job Description Job Description Salary: Who ar...      Software Developer   

                 created   latitude  longitude country       region  \
0   2025-06-15T03:29:02Z  43.706102 -79.402726  Canada      Ontario   
1   2025-05-20T17:15:34Z  43.706102 -79.402726  Canada      Ontario   
2   2025-04-20T22:56:29Z  45.169472 -76.730824  Canada      Ontario   
3   2025-05-31T22:30:19Z  47.029327 -94.640183  Canada         None   
4   2025-05-30T23:28:35Z  47.029327 -94.640183  Canada         None   
..                   ...        ...        ...     ...          ...   
95  2025-05-18T10:13:46Z  44.681555 -63.556363  Canada  Nova Scotia   
96  2025-06-07T08:50:55Z  45.466091 -75.704464  Canada       Québec   
97  2024-09-20T07:53:38Z  42.933380 -81.205550  Canada      Ontario   
98  2025-05-24T10:22:58Z  47.029327 -94.640183  Canada         None   
99  2025-04-24T15:20:44Z  42.981065 -81.248057  Canada      Ontario   

              city                   company_name  city_id  
0          Toronto                        Akkodis     17.0  
1          Toronto                        Aquanow     17.0  
2      Peel region      Tata Consultancy Services     14.0  
3             None                   iTech US Inc     11.0  
4             None  Critical Control Technologies     11.0  
..             ...                            ...      ...  
95  Halifax region                     Fundmetric      5.0  
96       Outaouais                   Pivot + Edge     13.0  
97       Middlesex             Proax Technologies      9.0  
98            None                   Nokia Global     11.0  
99       Middlesex             Proax Technologies      9.0  

[100 rows x 10 columns]

For cleanliness, remove city_id

In [16]:
employment_all = employment_all.drop(['city_id'], axis=1)

In [17]:
employment_all

description                   title  \
0   Akkodis is currently seeking a skilled Power B...            Data Analyst   
1   About us Aquanow is a trading and technology c...            Data Analyst   
2   Inclusion without Exception Tata Consultancy S...            Data Analyst   
3   Position: Technical Data Analyst Location: Mon...  Technical Data Analyst   
4   We are seeking an analytical and detail-orient...            Data Analyst   
..                                                ...                     ...   
95  Overview Fundmetric is looking for a Software ...      Software Developer   
96  Job Introduction Frequentis Group with headqua...      Software Developer   
97  Who are we? Proax Technologies is a leading Te...      Software Developer   
98  Description We are seeking a Software Develope...      Software Developer   
99  Job Description Job Description Salary: Who ar...      Software Developer   

                 created   latitude  longitude country       region  \
0   2025-06-15T03:29:02Z  43.706102 -79.402726  Canada      Ontario   
1   2025-05-20T17:15:34Z  43.706102 -79.402726  Canada      Ontario   
2   2025-04-20T22:56:29Z  45.169472 -76.730824  Canada      Ontario   
3   2025-05-31T22:30:19Z  47.029327 -94.640183  Canada         None   
4   2025-05-30T23:28:35Z  47.029327 -94.640183  Canada         None   
..                   ...        ...        ...     ...          ...   
95  2025-05-18T10:13:46Z  44.681555 -63.556363  Canada  Nova Scotia   
96  2025-06-07T08:50:55Z  45.466091 -75.704464  Canada       Québec   
97  2024-09-20T07:53:38Z  42.933380 -81.205550  Canada      Ontario   
98  2025-05-24T10:22:58Z  47.029327 -94.640183  Canada         None   
99  2025-04-24T15:20:44Z  42.981065 -81.248057  Canada      Ontario   

              city                   company_name  
0          Toronto                        Akkodis  
1          Toronto                        Aquanow  
2      Peel region      Tata Consultancy Services  
3             None                   iTech US Inc  
4             None  Critical Control Technologies  
..             ...                            ...  
95  Halifax region                     Fundmetric  
96       Outaouais                   Pivot + Edge  
97       Middlesex             Proax Technologies  
98            None                   Nokia Global  
99       Middlesex             Proax Technologies  

[100 rows x 9 columns]

Let us map the coordinates using ArcGis

In [18]:
import arcgis
from arcgis.gis import GIS
# Create a GIS object, as an anonymous user for this example
gis = GIS()

In [19]:
import sys
import arcgis
import notebook
import ipywidgets
print("Python:", sys.version)
print("Executable:", sys.executable)
print("ArcGIS:", arcgis.__version__)
print("Notebook:", notebook.__version__)
print("ipywidgets:", ipywidgets.__version__)

Python: 3.9.23 | packaged by conda-forge | (main, Jun  4 2025, 17:57:12) 
[GCC 13.3.0]
Executable: /home/edgar/miniconda3/envs/arcgis_env/bin/python
ArcGIS: 2.0.1
Notebook: 6.5.4
ipywidgets: 7.7.2


In [20]:
from arcgis.features import GeoAccessor, GeoSeriesAccessor
from arcgis.widgets import MapView

sdf = GeoAccessor.from_xy(employment_all, x_column='longitude', y_column='latitude')

map_view = gis.map("Canada", 4)

sdf['popup_info'] = sdf['title']
map_view.add_layer(sdf, options={"popupTemplate": {"title": "{popup_info}"}})


In [21]:
map_view

MapView(layout=Layout(height='400px', width='100%'))

As can be seen in the map above, most job openings in Canada are located in the Laurentian corridor(Which is where there is a higher population density)

It is important to note that this notebook can only run with the proper kernel installations:

Terminal:
cd ~/workspace/Job\ Portfolio/EdgarPortfolioDataScience
nano arcgis_env.yml

Yaml File:
name: arcgis_env
channels:
  - esri
  - conda-forge
  - defaults
dependencies:
  - python=3.9
  - arcgis=2.0.1
  - notebook=6.5.4
  - ipykernel=6.25.2
  - ipywidgets=7.7.2
  - jupyterlab=3.6.5
  - pandas
  - matplotlib
  - scikit-learn
  - nodejs=18

Terminal:
conda env create -f arcgis_env.yml
conda activate arcgis_env
pip install "urllib3<2.0"
pip install geomet
pip install ipywidgets==7.7.2 widgetsnbextension==3.6.4
jupyter nbextension enable --py widgetsnbextension --sys-prefix
jupyter nbextension enable --py arcgis --sys-prefix

and of course:
jupyter notebook